sw/device/tests/sim_dv/pwrmgr_normal_sleep_all_wake_ups.c

```diff
@@ -0,0 +1,345 @@
+  usbdev_set_wake_module_active(true);
+  usbdev_force_dx_pullup(kDpSel, true);
+  usbdev_force_dx_pullup(kDnSel, false);
+
+  // give the hardware a chance to recognize the wakeup values are the same
+  busy_spin_micros(20);  // 20us
+}
+
+/**
+ * aon timer config for test #5
+ * set wakeup signal in 50us
+ */
+static void prgm_aontimer_wakeup(void *dif) {
+  aon_timer_testutils_wakeup_config(dif, 10);
+}
+
+/**
+ * sensor ctrl config for test #6
+ * setup event trigger0
+ */
+static void prgm_sensor_ctrl_wakeup(void *dif) {
+  CHECK_DIF_OK(
+      dif_sensor_ctrl_set_ast_event_trigger(dif, 0, kDifToggleEnabled));
+}
+static const test_wakeup_sources_t kTestWakeupSources[] = {
+    {
+        .name = "SYSRST_CTRL",
+        .dif_handle = &sysrst_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceOne,
+        .config = prgm_sysrst_ctrl_wakeup,
+    },
+    {
+        .name = "ADC_CTRL",
+        .dif_handle = &adc_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceTwo,
+        .config = prgm_adc_ctrl_wakeup,
+    },
+    {
+        .name = "PINMUX",
+        .dif_handle = &pinmux,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceThree,
+        .config = prgm_pinmux_wakeup,
+    },
+    {
+        .name = "USB",
+        .dif_handle = &usbdev,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceFour,
+        .config = prgm_usb_wakeup,
+    },
+    {
+        .name = "AONTIMER",
+        .dif_handle = &aon_timer,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceFive,
+        .config = prgm_aontimer_wakeup,
+    },
+    {
+        .name = "SENSOR_CTRL",
+        .dif_handle = &sensor_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceSix,
+        .config = prgm_sensor_ctrl_wakeup,
+    },
+};
+
+/**
+ * External interrupt handler.
+ */
+void ottf_external_isr(void) {
+  dif_pwrmgr_irq_t irq_id;
+  top_earlgrey_plic_peripheral_t peripheral;
+
+  isr_testutils_pwrmgr_isr(plic_ctx, pwrmgr_isr_ctx, &peripheral, &irq_id);
+
+  // Check that both the peripheral and the irq id is correct
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralPwrmgrAon,
+        "IRQ peripheral: %d is incorrect", peripheral);
+  CHECK(irq_id == kDifPwrmgrIrqWakeup, "IRQ ID: %d is incorrect", irq_id);
+}
+
+static void execute_test(uint32_t wakeup_source) {
+  // Configure wakeup device
+  kTestWakeupSources[wakeup_source].config(
+      kTestWakeupSources[wakeup_source].dif_handle);
+  // Normal sleep
+  pwrmgr_testutils_enable_low_power(
+      &pwrmgr, kTestWakeupSources[wakeup_source].wakeup_src,
+      kDifPwrmgrDomainOptionMainPowerInLowPower);
+  LOG_INFO("Issue WFI to enter sleep %d", wakeup_source);
+  wait_for_interrupt();
+}
+
+/**
+ * Clean up pwrmgr wakeup reason register for the next round
+ */
+static void delay_n_clear(uint32_t delay_in_us) {
+  busy_spin_micros(delay_in_us);
+  CHECK_DIF_OK(dif_pwrmgr_wakeup_reason_clear(&pwrmgr));
+}
+
+/**
+ * Clean up wakup sources
+ */
+static void cleanup(uint32_t test_idx) {
```

Comment:
Rather than having explicit wait cycles, could we for example spin wait until the rv_plic has nothing pending? Otherwise this could fail if there are RTL changes.

Reply:
I rewrote cleanup parts (line 306), s.t.
polling pwrmgr wakeup status until it gets clean with in 30us.